### DISCUSSION TITLE
[Feat]: Add data attributes plugin

### DISCUSSION THREAD

**Discussion by DanielSusanu** [davidjerleke/embla-carousel#1051] [Ideas]
### Which variants of Embla Carousel are relevant to this feature request?

- [ ] embla-carousel **(Core)**
- [ ] embla-carousel-react
- [ ] embla-carousel-vue
- [ ] embla-carousel-svelte
- [ ] embla-carousel-autoplay
- [ ] embla-carousel-auto-scroll
- [ ] embla-carousel-solid
- [ ] embla-carousel-auto-height
- [X] embla-carousel-class-names
- [ ] embla-carousel-fade
- [ ] embla-carousel-docs **(Documentation)**
- [ ] embla-carousel-docs **(Generator)**

### Feature description

I am proposing the addition of a DataAttributes plugin that dynamically adds and removes custom HTML data attributes on the carousel and its slides based on various states and events. This would work similarly to the existing ClassNames plugin but would manipulate data attributes (data-*) instead of CSS classes.

### Additional Context

The DataAttributes plugin would:

1. Add or remove custom data attributes on the root element and individual slides.
2. Use attributes to indicate states such as data-dragging, data-snapped, and data-in-view.
3. Listen for key carousel events such as select, pointerDown, pointerUp, and slidesInView to manage these attributes dynamically.

I will be happy to contribute or fully implement this plugin.

Thank you for considering this feature request. I believe it will be a valuable addition to the ecosystem and enhance the way the carousel is integrate with modern CSS frameworks. 

### Before submitting

- [X] I've made research efforts and searched the documentation
- [X] I've searched for existing issues
- [X] I agree to follow this project's Contributing Guidelines for feature requests

**Comment by davidjerleke** [comment]
Hi @DanielSusanu,

I think this is a good idea! But at the time of writing, I'm not sure how to solve this in an optimal way. The `ClassNames` plugin has a lot of logic ready that can be used but I haven't thought about if it makes sense to create a more flexible plugin that solves both attribute and class name toggling, or if it's just better to duplicate the plugin and change the class attribute to data attributes.

Feel free to share your thoughts.

**Comment by DanielSusanu** [comment]
Hi @davidjerleke,

Thanks for considering this feature.

I believe keeping these as two separate plugins is the best approach, as it aligns with the single-purpose design approach of the existing plugins in the library. Combining the two functionalities would lead to a more complex codebase and would likely require renaming the plugin to describe both purposes.

Creating an additional DataAttributes plugin that initially shares most of the logic with the ClassNames plugin will be more straightforward. This will result in clearer, more maintainable code that focuses solely on data attribute handling.

This approach allows developers to easily choose between using class names, data attributes, or both, based on their needs.

Let me know what you think.

**Reply by davidjerleke** [reply]
@DanielSusanu thanks for sharing your thoughts.

> Creating an additional DataAttributes plugin that initially shares most of the logic with the ClassNames plugin will be more straightforward. This will result in clearer, more maintainable code that focuses solely on data attribute handling.

We should think about how we handle the attributes because maybe we wan't an option so devs can configure the `prefix`? I'm not sure about the use case at the time of writing but maybe some frameworks or similar is using custom attributes. For example:

```tsx
const options = { loop: false };

const plugins = [HtmlAttributes({ 
  prefix: 'data-', // Default, so can be omitted but here just for clarity
  snapped: 'is-snapped',
})];

const emblaApi = EmblaCarousel(emblaNode, options, plugins);
```

For snapped slides the result would be:

```html
<div class="embla__slide" data-is-snapped="true">
```

For slides that aren't snapped the result would be:

```html
<div class="embla__slide" data-is-snapped="false">
```

---

And changing the prefix would make it work like this:

```tsx
const options = { loop: false };

const plugins = [HtmlAttributes({ 
  prefix: '', // Let's leave it empty
  snapped: 'is-snapped',
})];

const emblaApi = EmblaCarousel(emblaNode, options, plugins);
```

For snapped slides the result would be:

```html
<div class="embla__slide" is-snapped="true">
```

For slides that aren't snapped the result would be:

```html
<div class="embla__slide" is-snapped="false">
```

Any thoughts about this idea?

**Reply by DanielSusanu** [reply]
Hi @davidjerleke 

In regard to replacing the data- prefix with a different prefix, I don't think it's the best idea due to the following reasons:

1. JavaScript Compatibility: Custom attributes without the data- prefix won't be accessible through the dataset property in JavaScript. This is also a hint that the best practice for custom attributes is to use the data- prefix.
2. Potential Conflicts: Custom prefixes may compete with built-in HTML attribute names, leading to potential naming conflicts.
3. HTML Validation: Some HTML validators might flag non-standard attributes as errors, which could impact code validation and quality checks.
4. I assume most developers would expect custom attributes to be prefixed with data- for consistency, could be wrong.

That said, it depends on how flexible we want the plugin to be or if we want to enforce best practices.

The ability to configure the attribute name itself is a good idea.

Let me know what you think.

**Reply by davidjerleke** [reply]
@DanielSusanu let's skip the prefix for now and name the plugin `DataAttributes` and copy the ClassNames logic but apply data-attributes instead 🙂.

How should we apply the data attributes? Let's take the is snapped as an example. Should the attributtes be present even when they're `false` like this:

Snapped:

```html
<div class="embla__slide" data-is-snapped="true">
```

NOT snapped:

```html
<div class="embla__slide" data-is-snapped="false">
```

...or should the attribute be absent?


Snapped:

```html
<div class="embla__slide" data-is-snapped="true">

<div class="embla__slide" data-is-snapped> <!-- Or this? -->
```

NOT snapped:

```html
<div class="embla__slide" >
```

**Reply by DanielSusanu** [reply]
@davidjerleke I've had a look at how attributes are handled by other libraries like `radix-ui` & `headlessui`.

Both of the use the `data-` prefix. Headlessui also add 'headlessui' after 'data-', all their attributes have the prefix `data-headlessui-`

Radix:
1. Uses data- prefixed attributes, similar to your first example.
2. Attributes are always present with a boolean value.

>Snapped:
>
>```
><div class="embla__slide" data-is-snapped="true">
>```
>NOT snapped:
>```
><div class="embla__slide" data-is-snapped="false">
>```

Headlessui:

1. They have an attribute called `data-headlessui-state`, that is always present. This contains a string list of all active states but is not used for styling.
2. Then for each state they add and remove the attribute. When the attribute is present they don't add the `true` value.

None of the approaches are wrong, I think I prefer better the attribute being absent. And when is present to not have a value. I also like the idea to include the library name in the attribute, will prevent potential conflicts and makes the attribute easier to identify as related to Embla library. 

Snapped:

```
<div class="embla__slide" data-embla-is-snapped >
```

NOT snapped:


```
<div class="embla__slide" >
```

Let me know your thoughts!

**Reply by davidjerleke** [reply]
@DanielSusanu ok so maybe we should go for this:

### A prefix option that **adds** to the core `data-` prefix? 

Example with default option:

```js
const plugins = [HtmlAttributes({ 
  prefix: 'embla', // Default is embla
  snapped: 'is-snapped',
})];
```

Which gives us:

```html
<div class="embla__slide" data-embla-is-snapped>
```

However setting it to something:

```js
const plugins = [HtmlAttributes({ 
  prefix: 'headlessui',
  snapped: 'is-snapped',
})];
```

Gives us this instead:

```html
<div class="embla__slide" data-headlessui-is-snapped>
```

---

### An option to apply data-attributes even when false:

```js
const plugins = [HtmlAttributes({ 
  applyAttributeWhenFalse: false, // Default
  snapped: 'is-snapped',
})];
```

Which gives us:

```html
<div class="embla__slide" data-embla-is-snapped>
```

However setting it to `true`:

```js
const plugins = [HtmlAttributes({ 
  applyAttributeWhenFalse: true,
  snapped: 'is-snapped',
})];
```

Which gives us:

```html
<div class="embla__slide" data-embla-is-snapped="true">
```

```html
<div class="embla__slide" data-embla-is-snapped="false">
```

---

What do you think?

**Reply by DanielSusanu** [reply]
@davidjerleke Sounds great. 

Lets go for it.

**Reply by davidjerleke** [reply]
If anyone else other than me @davidjerleke and @DanielSusanu has any input to share, please do it here as soon as possible before the implementation of this plugin starts. Otherwise breaking changes will be postponed to the next major release.

**Reply by davidjerleke** [reply]
@DanielSusanu I want to ask if you would want to develop the `DataAttributes` plugin and create a PR?

If yes, it's good for you to know that I will make some improvements to the `ClassNames` plugin when I get the chance that might be relevant to use in the `DataAttributes` plugin. So you can either wait until I'm done with that or start working on it now with the knowledge that some changes may be requested during the development process.

**Reply by DanielSusanu** [reply]
Hi @davidjerleke yes, I can develop the plugin.

I will wait till you've done your updates.
Do you have a timeline in your mind?

**Reply by davidjerleke** [reply]
@DanielSusanu that's good news 👍. 

> Do you have a timeline in your mind?

Unfortunately no. My spare time is very limited and I can't really plan when I will get my next chance to work on this project. So can't make promises or give ETA's.

**Reply by DanielSusanu** [reply]
@davidjerleke No worries. 
Will catch up when you have time to work on this.

**Reply by davidjerleke** [reply]
Hi again @DanielSusanu,

I just made the improvements I mentioned to the `ClassNames` plugin. Here's [the current state of the plugin](https://github.com/davidjerleke/embla-carousel/blob/master/packages/embla-carousel-class-names/src/components/ClassNames.ts) with the updates if you want to use that as a base for the `DataAttributes` plugin.

Best,
David

**Reply by davidjerleke** [reply]
@DanielSusanu just want to double check that you've seen my [last comment](https://github.com/davidjerleke/embla-carousel/discussions/1051#discussioncomment-11279855). Not to stress you though.

**Reply by DanielSusanu** [reply]
Hey @davidjerleke, I've seen your message but not your changes. Will do this coming weekend, and get in touch again.

**Reply by davidjerleke** [reply]
@DanielSusanu maybe I'm misunderstanding you then 🙂? This is the PR with the improvements to the `ClassNames` plugin I had in mind:

- #1061

My point was that I wanted to do these improvements first so you can use that base (including the improvements) when you begin working on the `DataAttributes` plugin. [This is](https://github.com/davidjerleke/embla-carousel/blob/master/packages/embla-carousel-class-names/src/components/ClassNames.ts) how the `ClassNames` plugin code looks after I recently merged the PR above. 

So now that I'm done feel free to start working on the `DataAttributes` plugin. If you have any questions about Embla let me know and I will answer you when I get the chance.
